### ISSUE TITLE
Stabilize guaranteed compile time evaluation of unnamed constant items

### ISSUE THREAD

**Issue by XrXr** [rust-lang/rust#93838]
This is a request to stabilize the implementation fact that the definition of [unnamed constant items] are always evaluated at compile time (through [CTFE]). There is no associated RFC as this feature is already available in stable releases and this is mostly a request for using an [FCP] to reach consensus for future support.

The [stabilization][const panic stabilization] of `const-panic` introduced a way for CTFE to emit hard compile errors. This allows users to perform compile time assertions using familiar language constructs and a growing set of `const` APIs. Before `const-panic`, users could emit the `const_err` lint, which was less dependable since whether lints fail builds is configurable.

Authors of compile time assertions that utilize `const-panic` would probably like to have confidence that what they wrote is in fact evaluated by `rustc` during compilation. Currently, users can ensure evaluation by utilizing const contexts that are inputs to the type checker:

```rust
const _: [(); panic!("custom error")] = [];
```

This trick relies on the fact that type checking is guaranteed to happen during compilation. Utilizing this to initiate constant evaluation feels more cumbersome than necessary.

This report aims to stabilize a more obvious way to initiate compile time evaluation.

## What is being stabilized

The current `rustc` behavior that unnamed constant definitions are evaluated at least once at compile time. Note that compile time evaluation happens even when the function enclosing the unnamed constant is unused:

```rust
fn unused() {
    const _: () = panic!("always a compile error");
}
```

Stabilizing this seems unlikely to cause maintenance issues in the future due to the constrained nature of unnamed constants. However, I defer to people with `rustc` maintenance experience to judge this claim.

## What is _not_ being stabilized

The evaluation timing of named constant items. Named constant items can be associated constants, which could use generics. Unnamed constants [cannot be associated constants][unnamed constant items] and [cannot use generic parameters][outer func error playground]. While it might be useful to also stabilize the evaluation timing of named constant items, they require more consideration and are out of scope for this report.

---

I would like to thank Ralf Jung, Oli Scherer, and Eric Huss for the generous amount of guidance I received. Here is the Zulip [discussion][zulip] leading up to this stabilization request.

[const panic stabilization]: https://blog.rust-lang.org/2021/12/02/Rust-1.57.0.html#panic-in-const-contexts
[unnamed constant items]: https://doc.rust-lang.org/stable/reference/items/constant-items.html#unnamed-constant
[constant evaluation]: https://doc.rust-lang.org/stable/reference/const_eval.html
[CTFE]: https://github.com/rust-lang/rust/issues/11621
[outer func error playground]: https://gist.github.com/XrXr/0336f977dd17d490d895a721ffc61207
[zulip]: https://rust-lang.zulipchat.com/#narrow/stream/213817-t-lang/topic/Is.20a.20FCP.20required.20for.20basic.20CTFE.20guarantees.3F/near/269816535
[FCP]: https://rustc-dev-guide.rust-lang.org/stabilization_guide.html#fcp

**Comment by joshtriplett** [comment]
We discussed this in today's @rust-lang/lang meeting. We'd be fine with stabilizing such a guarantee.

We'd want to make sure that the language used to make the guarantee still allows the compiler to optimize away code (e.g. anything that doesn't have a visible side effect such as a panic).

**Comment by joshtriplett** [comment]
Do we already have tests that confirm this behavior?

**Comment by pnkfelix** [comment]
(just passing this on to the compiler team to double-check that people there are okay with what's being proposed here. And if firmer language is required to be checked against the rust-lang/reference, then we need to check that nominations there show up on the T-compiler agenda.)

**Comment by apiraino** [comment]
Discussed in T-compiler meeting ([on Zulip](https://zulip-archive.rust-lang.org/stream/238009-t-compiler/meetings/topic/.5Bweekly.5D.202022-03-03.20.2354818.html#273991518))

@rustbot label -I-compiler-nominated

**Comment by Aaron1011** [comment]
> Note that compile time evaluation happens even when the function enclosing the unnamed constant is unused:

I think we should clarify that this is true even for unused *generic* functions:

```rust
fn foo<T>() {
    const _: () = {
        panic!("Const eval!");
    };
}
```

```
error[[E0080]](https://doc.rust-lang.org/nightly/error-index.html#E0080): evaluation of constant value failed
 [--> src/lib.rs:3:9
](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021#)  |
3 |         panic!("Const eval!");
  |         ^^^^^^^^^^^^^^^^^^^^^ the evaluated program panicked at 'Const eval!', src/lib.rs:3:9
  |
  = note: this error originates in the macro `$crate::panic::panic_2021` (in Nightly builds, run with -Z macro-backtrace for more info)
```

Since (as @XrXr mentioned) unnamed const items cannot use generics, we evaluate unnamed const items in the body of unused generic functions.

**Comment by MauriceKayser** [comment]
I think this is related:
```rust
struct A();
impl A {
    const _1: [(); panic!("1")] = [];
    const _2: () = panic!("2");
}
```
Currently `1` panics, but `2` does not.

**Comment by XrXr** [comment]
I submitted PR to the reference https://github.com/rust-lang/reference/pull/1328 to document the timing.

> We'd want to make sure that the language used to make the guarantee still allows the compiler to optimize away code (e.g. anything that doesn't have a visible side effect such as a panic).

@joshtriplett can I ask for a review of the language?

**Comment by XrXr** [comment]
@MauriceKayser all of those are named constant items so is out of scope for this issue. I also find it odd that associated constants are not always evaluated. From my understanding the timing there is related to how `impl` blocks can be generic, so even though in this example it could be evaluated, it is not, as in general evaluation is not possible until the generic variables are concrete. This gives a sort of consistency, even though it's a bit surprising to me.

In any case, it's not in-scope for this issue.

**Comment by XrXr** [comment]
> Do we already have tests that confirm this behavior?

The only related test I found is [`const-eval/unused-broken-const.rs`](https://github.com/rust-lang/rust/tree/master/tests/ui/consts/const-eval/unused-broken-const.rs), first [introduced in 2018](https://github.com/rust-lang/rust/blob/7ed7fc881e7f4dabcdc45646257ddb979822d7f4/src/test/ui/consts/const-eval/unused-broken-const.rs). It doesn't test `const _` specifically and doesn't test the unused generic function case mentioned above. I guess the discussed cases are covered because `UnusedBrokenConst: LateLintPass`.

There are many tests that implicitly depend on the definition being evaluated, though. Some examples:
 - [const-eval/const_panic.rs](https://github.com/rust-lang/rust/blob/1.67.1/src/test/ui/consts/const-eval/const_panic.rs)
 - [const-eval/const_raw_ptr_ops.rs](https://github.com/rust-lang/rust/blob/1.67.1/src/test/ui/consts/const-eval/const_raw_ptr_ops.rs)
 - [const-eval/const_raw_ptr_ops2.rs](https://github.com/rust-lang/rust/blob/1.67.1/src/test/ui/consts/const-eval/const_raw_ptr_ops2.rs)
 - [const-eval/dangling.rs](https://github.com/rust-lang/rust/blob/1.67.1/src/test/ui/consts/const-eval/dangling.rs)

Should I expand `const-eval/unused-broken-const.rs`?

**Comment by XrXr** [comment]
Now that https://github.com/rust-lang/reference/pull/1328 is merged, I'm going to go ahead and close this. This functionality is probably stable enough even without any explicit confirmation through an FCP, RFC or what have you. Weakening this guarantee would probably break a lot of crates in the ecosystem, so I trust that such change will not happen without careful consideration.

Thanks everyone for your attention!
